### DISCUSSION TITLE
Feature: Support unicode character types

### DISCUSSION THREAD

**Discussion by gracicot** [nlohmann/json#2337] [Ideas]
This is a feature request. I wondered if this is possible. The JSON format is UTF-8. C++11 added `char16_t` and `char32_t` for unicode characters, and lately C++20 added `char8_t`.

Since JSON is specified as UTF-8, this library should at least support  `char8_t`. This add more safety since we can validate and check the character encoding at compile time, prohibiting sending strings to functions that expects another encoding.

#### What is the issue you have?

Right now, trying to use `char8_t` with `std::u8string` in a `basic_json` result in error.

#### Please describe the steps to reproduce the issue.

```c++
using u8json = basic_json<std::map, std::vector, std::u8string>;
u8json value = {};
```

[Compiler explorer link](https://godbolt.org/z/zx18PY)

#### What is the expected behavior?

The `u8json` alias should work, and conversions to `std::u8string` and `std::u8string_view` should work instead of `std::string` using this alias.

```c++
auto json = u8json{
    {u8"key", u8"value"}
};

auto key = std::u8string{json[u8"key"]};
```

#### And what is the actual behavior instead?

A compilation error.

#### Which compiler and operating system are you using?

- Compiler: GCC (trunk)
- Operating system: Any

#### Which version of the library did you use?

- [X] latest release version 3.9.0
- [ ] other release - please state the version: ___
- [ ] the `develop` branch

#### If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

- [X] yes
- [ ] no - please copy/paste the error message below

**Comment by nlohmann** [comment]
This has been substantially addressed since: include/nlohmann/json.hpp now provides a `char8_t`-based `operator""_json` (guarded by `__cpp_char8_t`), and `from_json`/`to_json` have explicit `char8_t`/`std::u8string_view` conversion paths guarded by `__cpp_lib_char8_t`. Using `std::u8string` directly as a custom `StringType` alias may still have rough edges, but native `char8_t` string literal and conversion support — the core of this request — is now in place.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*

**Reply by gracicot** [reply]
Thank you so much!
